crates/codegen/schema/src/validation/rules/references.rs

```diff
@@ -34,65 +37,102 @@ impl<'v> ReferencesCollector<'v> {
             references: HashSet::new(),
         };
     }
+
+    fn assert_is_scanner_reference(
```

Comment:
Love these hlpers!
A couple of nits:

- I suggest `check_*` instead of `assert_is_*` since it is not actually asserting/panicing, but also modifying the state (collecting references).
- I suggest moving these helpers to the `impl` block that has `check_unused`, after the parsers, to maintain how this file is ordered for readability. Entry point, then visitor, then validation, then errors.

Reply:
done